### PR TITLE
support argmax/argmin without dim kwargs and fix adaptive_max_pool3d

### PR DESCRIPTION
**Pull request by cccclai**
Summary: As title, in PyTorch, when dim is not set, it will flatten the input and get argmax as dim=0. Add a pass to reshape the input when dim is not set and consolidate test case

edit:
1. Apply to argmin too
2. Add `exir_ops.edge.aten.adaptive_max_pool3d.default` to the to be implemented op list to pass the error

Differential Revision: D83606497

### CODE REVIEW

File: backends/qualcomm/_passes/insert_reshape_for_argmax.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# All rights reserved
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from executorch.exir.pass_base import ExportPass, PassResult
+from executorch.exir.passes import dead_code_elimination_pass
+
+class InsertReshapeForArgmax(ExportPass):
```
Comment (haowhsu-quic) [inline backends/qualcomm/_passes/insert_reshape_for_argmax.py]: I think we might have other ops like `argmin` which can leverage this pass as well. Do you mind rewording the class name into more generic one? Something like `InsertReshapeForReduceOp`.

File: backends/qualcomm/_passes/insert_reshape_for_argmax.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    def __init__(self):
+        super().__init__()
+        self.op_map = {torch.ops.aten.argmax.default}
+
+    def call(self, graph_module: torch.fx.GraphModule):
+        graph = graph_module.graph
+        modified = False
+
+        for n in list(graph.nodes):
```
Comment (haowhsu-quic) [inline backends/qualcomm/_passes/insert_reshape_for_argmax.py]: I think we can use the iterator directly: `for n in graph.nodes:`.

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                dim_arg = None if len(n.args) == 1 else n.args[1]
+
+                if dim_arg is None:
+                    inp = n.args[0]
+
+                    # Insert reshape before argmax
+                    with graph.inserting_before(n):
+                        reshape_node = graph.create_node(
+                            "call_function",
+                            torch.ops.aten.reshape.default,
```
Comment (haowhsu-quic) [inline backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py:42]: I think we use `torch` namespace for graph which needs decomposition before `prepare_pt2e` and `exir_ops` namespace for floating point graph which goes directly to `to_edge_transform_and_lower` API.

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                            {},
+                        )
+                        reshape_node.meta = dict(inp.meta)
+                        if "val" in inp.meta:
+                            reshape_node.meta["val"] = inp.meta["val"].reshape(-1)
+
+                    # Rewrite argmax: take reshape_node as input, set dim=0
+                    n.args = (reshape_node, 0, *n.args[2:])
+                
+                modified = True
```
Comment (haowhsu-quic) [inline backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py]: I think you need one more indention to have this statement fall into branch.

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py (hunk 3 above)
Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py:42]: I see, do you mean I need to use the op with different namespace here?

File: backends/qualcomm/tests/test_passes.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import unittest
+import torch
+from executorch.backends.qualcomm._passes import InsertReshapeForArgmax
+
+class TestPasses(unittest.TestCase):
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_passes.py:7]: Thank you for adding this file, it's helpful for us to test all the passes thoroughly.

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py (hunk 3 above)
Comment (haowhsu-quic) [reply backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py:42]: It's a convention we're using but I guess the op builder could still work since it only recognizes part of op's target string.

Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py:42]: I was actually using it for both `transform_for_annotation_pipeline` and `transform_for_export_pipeline`. I was thinking it could be a good idea to have it in preprocess, but preprocess happens after partition, and op_argmax.py will run during partition as well. I think in this case, we need to have it similar to those decompose passes, which means we use `torch` namespace

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py (hunk 4 above)
Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py]: updated

File: backends/qualcomm/_passes/insert_reshape_for_argmax.py (hunk 2 above)
Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_argmax.py]: updated

File: backends/qualcomm/_passes/insert_reshape_for_argmax.py (hunk 1 above)
Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_argmax.py]: updated

Comment (winskuo-quic) [reply backends/qualcomm/_passes/insert_reshape_for_argmax.py]: I think the new commit wasn't pushed successfully. Could you try again? Thanks

Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_argmax.py]: oh sorry, just push

File: backends/qualcomm/builders/op_argmax.py [hunk 6]
```diff
@@ -46,7 +46,7 @@ def define_node(
             nodes_to_wrappers,
         )
         argmax_output_tensors = [argmax_out_tensor_wrapper]
-
+        print("argmax node.args: ", node.args)
```
Comment (mergennachin) [inline backends/qualcomm/builders/op_argmax.py]: remove

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) Qualcomm Innovation Center, Inc.
```
Comment (mergennachin) [inline backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py]: You, representing Meta, contributed this file, right? If so, add Meta copyright

File: backends/qualcomm/builders/op_argmax.py (hunk 6 above)
Comment (cccclai) [reply backends/qualcomm/builders/op_argmax.py]: updated

File: backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py (hunk 7 above)
Comment (cccclai) [reply backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py]: updated

**Comment by cccclai** [comment]
Can I get another round of review on this? I fix a few more failing tests including `argmin`, and `adaptive_max_pool3d`

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 8]
```diff
@@ -173,14 +173,60 @@ def test_qnn_backend_arange(self):
                 self.lower_module_and_test_output(module, sample_input)
 
     def test_qnn_backend_argmax(self):
-        module = Argmax()  # noqa: F405
-        sample_input = (torch.randn(16, 3, 4, 4),)
-        self.lower_module_and_test_output(module, sample_input)
+        test_cases = [
+            {
+                "module": Argmax(),  # noqa: F405
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: I think we have keywords such as "QCOM_MODULE", "QCOM_SAMPLE_INPUTS". It would be great if you could use those.

Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: Yeah let me update

**Review by haowhsu-quic — APPROVED** [review]
Thank you!

File: backends/qualcomm/partition/common_defs.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 to_be_implemented_operator = [
     exir_ops.edge.aten._adaptive_avg_pool3d.default,
     exir_ops.edge.aten.adaptive_max_pool2d.default,
+    exir_ops.edge.aten.adaptive_max_pool3d.default,
```
Comment (kimishpatel) [inline backends/qualcomm/partition/common_defs.py:20]: is this pr just about argmax? i see maxpool added here. if it is about reduce ops then please do update the title

**Review by kimishpatel — APPROVED** [review]

File: backends/qualcomm/partition/common_defs.py (hunk 9 above)
Comment (cccclai) [reply backends/qualcomm/partition/common_defs.py:20]: Yeah updated
